mlflow/gateway/providers/palm.py

```diff
@@ -0,0 +1,189 @@
+            **{
+                "candidates": [
+                    {
+                        "message": {
+                            "role": c["author"],
+                            "content": c["content"],
+                        },
+                        "metadata": {},
+                    }
+                    for c in resp["candidates"]
+                ],
+                "metadata": {
+                    "model": self.config.model.name,
+                    "route_type": self.config.route_type,
+                },
+            }
+        )
+
+    async def completions(self, payload: completions.RequestPayload) -> completions.ResponsePayload:
+        payload = jsonable_encoder(payload, exclude_none=True)
+        self.check_for_model_field(payload)
+        key_mapping = {
+            "stop": "stopSequences",
+            "candidate_count": "candidateCount",
+            "max_tokens": "maxOutputTokens",
+        }
+        for k1, k2 in key_mapping.items():
+            if k2 in payload:
+                raise HTTPException(
+                    status_code=400, detail=f"Invalid parameter {k2}. Use {k1} instead."
+                )
+        payload = rename_payload_keys(payload, key_mapping)
+        payload["prompt"] = {"text": payload["prompt"]}
+        resp = await self._request(
+            f"{self.config.model.name}:generateText",
+            payload,
+        )
+        # Response example (https://developers.generativeai.google/api/rest/generativelanguage/models/generateText)
+        # ```
+        # {
+        #   "candidates": [
+        #     {
+        #       "output": "Once upon a time, there was a young girl named Lily...",
+        #       "safetyRatings": [
+        #         {
+        #           "category": "HARM_CATEGORY_DEROGATORY",
+        #           "probability": "NEGLIGIBLE"
+        #         }, ...
+        #       ]
+        #     {
+        #       "output": "Once upon a time, there was a young boy named Billy...",
+        #       "safetyRatings": [
+        #           ...
+        #       ]
+        #     }
+        #   ]
+        # }
+        # ```
+        return completions.ResponsePayload(
+            **{
+                "candidates": [
+                    {
+                        "text": c["output"],
+                        "metadata": {},
+                    }
+                    for c in resp["candidates"]
+                ],
+                "metadata": {
+                    "model": self.config.model.name,
+                    "route_type": self.config.route_type,
+                },
+            }
+        )
+
+    async def embeddings(self, payload: embeddings.RequestPayload) -> embeddings.ResponsePayload:
+        payload = jsonable_encoder(payload, exclude_none=True)
+        self.check_for_model_field(payload)
+
+        # Ensure 'text' is a string
+        if isinstance(payload["text"], list):
+            payload["text"] = ",".join(payload["text"])
```

Comment:
can we use https://developers.generativeai.google/api/rest/generativelanguage/models/batchEmbedText?

Reply:
Added!